src/libopensc/card-dnie.c

```diff
@@ -729,6 +729,62 @@ int dnie_match_card(struct sc_card *card)
 	LOG_FUNC_RETURN(card->ctx, result);
 }
 
+static int dnie_sm_free_wrapped_apdu(struct sc_card *card,
+		struct sc_apdu *plain, struct sc_apdu **sm_apdu)
+{
+	struct sc_context *ctx = card->ctx;
+	int rv = SC_SUCCESS;
+
+	LOG_FUNC_CALLED(ctx);
+	if (!sm_apdu)
+		LOG_FUNC_RETURN(ctx, SC_ERROR_INVALID_ARGUMENTS);
+	if (!(*sm_apdu))
+		LOG_FUNC_RETURN(ctx, SC_SUCCESS);
+
+	plain->resp = (*sm_apdu)->resp;
+	plain->resplen = (*sm_apdu)->resplen;
+	plain->sw1 = (*sm_apdu)->sw1;
+	plain->sw2 = (*sm_apdu)->sw2;
```

Comment:
OK, I understand. But why do you continue carrying this historic burden? This still leaves the code a little bit confusing.

 You could rewrite the code so that `dnie_sm_get_wrapped_apdu()` encrypts the APDU and `dnie_sm_free_wrapped_apdu()` decrypts the response. This allows removing the special case in `dnie_transmit_apdu()`, which checks for `SM_MODE_TRANSMIT`. This is already done via `sc_transmit_apdu()` (https://github.com/OpenSC/OpenSC/blob/master/src/libopensc/apdu.c#L391 to be exactly).